File: sdks/ts/memory/src/ingest/state-store.test.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Contract test suite for PipelineStateStore implementations.
 *
 * Both FilePipelineStateStore and PostgresPipelineStateStore must pass the
 * same behavioural contract. The file-based implementation is tested here
 * directly (no Docker). The Postgres implementation uses an in-memory fake
 * that honours the same interface.
 */

import { afterEach, beforeEach, describe, expect, it } from 'vitest'
import { createMemStore } from '../store/memstore.js'
import type { Store } from '../store/index.js'
import type { PipelineStateEntry, PipelineStateStore } from './state-store.js'
import { FilePipelineStateStore } from './state-store.js'
import { PostgresPipelineStateStore } from './state-store-pg.js'
import type { PgSql } from './state-store-pg.js'

const BRAIN_ID = 'test-brain'

/** Valid 64-character lowercase hex hash for tests. */
const DEFAULT_HASH = 'abc123def4560000000000000000000000000000000000000000000000000000'

const makeEntry = (overrides: Partial<PipelineStateEntry> = {}): PipelineStateEntry => ({
  documentHash: DEFAULT_HASH,
  brainId: BRAIN_ID,
  stage: 'stored',
  retryCount: 0,
  createdAt: new Date('2026-01-01T00:00:00Z'),
  updatedAt: new Date('2026-01-01T00:00:00Z'),
  ...overrides,
})

/**
 * In-memory fake that mimics the PostgresPipelineStateStore's SQL behaviour
 * without requiring a real database connection. Routes based on the query
 * string passed to `unsafe()`.
 */
const createFakePgSql = (): PgSql => {
  const rows = new Map<string, Record<string, unknown>>()

  const taggedTemplate = (
    _strings: TemplateStringsArray,
    ..._values: unknown[]
  ): Promise<ReadonlyArray<unknown>> => Promise.resolve([])

  const unsafe = (query: string, params?: unknown[]): Promise<ReadonlyArray<unknown>> => {
    const safeParams = params ?? []

    if (query.includes('SELECT') && query.includes('LIMIT 1')) {
      const documentHash = safeParams[0] as string
      const row = rows.get(documentHash)
      if (row === undefined) return Promise.resolve([])
      return Promise.resolve([row])
    }

    if (query.includes('SELECT') && query.includes('NOT IN')) {
      const brainId = safeParams[0] as string
      const matching: Record<string, unknown>[] = []
      for (const row of rows.values()) {
        if (
          row['brain_id'] === brainId &&
          row['stage'] !== 'completed' &&
          row['stage'] !== 'failed'
        ) {
          matching.push(row)
        }
      }
      matching.sort((a, b) => {
        const aTime = (a['created_at'] as Date).getTime()
        const bTime = (b['created_at'] as Date).getTime()
        return aTime - bTime
      })
      return Promise.resolve(matching)
    }

    if (query.includes('INSERT')) {
      const documentHash = safeParams[0] as string
      const row: Record<string, unknown> = {
        document_hash: safeParams[0],
        brain_id: safeParams[1],
        stage: safeParams[2],
        retry_count: safeParams[3],
        last_error: safeParams[4],
        created_at: safeParams[5],
        updated_at: safeParams[6],
        completed_at: safeParams[7],
      }
      rows.set(documentHash, row)
      return Promise.resolve([])
    }

    if (query.includes('DELETE')) {
      const documentHash = safeParams[0] as string
      rows.delete(documentHash)
      return Promise.resolve([])
    }

    return Promise.resolve([])
  }

  const begin = async <T>(fn: (s: PgSql) => Promise<T>): Promise<T> => fn(sql)

  const sql = Object.assign(taggedTemplate, { begin, unsafe }) as unknown as PgSql

  return sql
}

type StoreFactory = {
  readonly name: string
  readonly create: () => { stateStore: PipelineStateStore; cleanup: () => Promise<void> }
}

const factories: readonly StoreFactory[] = [
  {
    name: 'FilePipelineStateStore',
    create: () => {
      const store: Store = createMemStore()
      const stateStore = new FilePipelineStateStore(store)
      return { stateStore, cleanup: async () => store.close() }
    },
  },
  {
    name: 'PostgresPipelineStateStore',
    create: () => {
      const sql = createFakePgSql()
      const stateStore = new PostgresPipelineStateStore({ sql })
      return { stateStore, cleanup: async () => {} }
    },
  },
]

for (const factory of factories) {
  describe(`PipelineStateStore contract: ${factory.name}`, () => {
    let stateStore: PipelineStateStore
    let cleanup: () => Promise<void>

    beforeEach(() => {
      const created = factory.create()
      stateStore = created.stateStore
      cleanup = created.cleanup
    })

    afterEach(async () => {
      await cleanup()
    })

    it('returns undefined when no state exists', async () => {
      const entry = await stateStore.get('0000000000000000000000000000000000000000000000000000000000000000')
      expect(entry).toBeUndefined()
    })

    it('round-trips state through set and get', async () => {
      const entry = makeEntry()
      await stateStore.set(entry)
      const retrieved = await stateStore.get(entry.documentHash)
      expect(retrieved).toBeDefined()
      expect(retrieved!.documentHash).toBe(entry.documentHash)
      expect(retrieved!.brainId).toBe(entry.brainId)
      expect(retrieved!.stage).toBe(entry.stage)
      expect(retrieved!.retryCount).toBe(entry.retryCount)
      expect(retrieved!.createdAt.toISOString()).toBe(entry.createdAt.toISOString())
      expect(retrieved!.updatedAt.toISOString()).toBe(entry.updatedAt.toISOString())
    })

    it('overwrites existing entry on set', async () => {
      const entry = makeEntry({ stage: 'stored' })
      await stateStore.set(entry)

      const updated = makeEntry({
        stage: 'chunked',
        updatedAt: new Date('2026-01-02T00:00:00Z'),
      })
      await stateStore.set(updated)

      const retrieved = await stateStore.get(entry.documentHash)
      expect(retrieved).toBeDefined()
      expect(retrieved!.stage).toBe('chunked')
      expect(retrieved!.updatedAt.toISOString()).toBe('2026-01-02T00:00:00.000Z')
    })

    it('lists only incomplete entries', async () => {
      const received = makeEntry({ documentHash: '1111111111111111111111111111111111111111111111111111111111111111', stage: 'received' })
      const stored = makeEntry({ documentHash: '2222222222222222222222222222222222222222222222222222222222222222', stage: 'stored' })
      const completed = makeEntry({ documentHash: '3333333333333333333333333333333333333333333333333333333333333333', stage: 'completed' })
      const failed = makeEntry({ documentHash: '4444444444444444444444444444444444444444444444444444444444444444', stage: 'failed' })
      const embedded = makeEntry({ documentHash: '5555555555555555555555555555555555555555555555555555555555555555', stage: 'embedded' })

      await stateStore.set(received)
      await stateStore.set(stored)
      await stateStore.set(completed)
      await stateStore.set(failed)
      await stateStore.set(embedded)

      const incomplete = await stateStore.listIncomplete(BRAIN_ID)
      const hashes = incomplete.map((e) => e.documentHash)

      expect(hashes).toContain('1111111111111111111111111111111111111111111111111111111111111111')
      expect(hashes).toContain('2222222222222222222222222222222222222222222222222222222222222222')
      expect(hashes).toContain('5555555555555555555555555555555555555555555555555555555555555555')
      expect(hashes).not.toContain('3333333333333333333333333333333333333333333333333333333333333333')
      expect(hashes).not.toContain('4444444444444444444444444444444444444444444444444444444444444444')
      expect(incomplete).toHaveLength(3)
    })

    it('listIncomplete filters by brainId', async () => {
      const ours = makeEntry({ documentHash: 'aaaa000000000000000000000000000000000000000000000000000000000000', brainId: BRAIN_ID, stage: 'stored' })
      const theirs = makeEntry({
        documentHash: 'bbbb000000000000000000000000000000000000000000000000000000000000',
        brainId: 'other-brain',
        stage: 'stored',
      })

      await stateStore.set(ours)
      await stateStore.set(theirs)

      const incomplete = await stateStore.listIncomplete(BRAIN_ID)
      expect(incomplete).toHaveLength(1)
      expect(incomplete[0]!.documentHash).toBe('aaaa000000000000000000000000000000000000000000000000000000000000')
    })

    it('deletes state by document hash', async () => {
      const entry = makeEntry()
      await stateStore.set(entry)

      await stateStore.delete(entry.documentHash)
      const retrieved = await stateStore.get(entry.documentHash)
      expect(retrieved).toBeUndefined()
    })

    it('delete on non-existent hash does not throw', async () => {
      await expect(stateStore.delete('0000000000000000000000000000000000000000000000000000000000000001')).resolves.toBeUndefined()
    })

    it('preserves lastError field', async () => {
      const entry = makeEntry({
        stage: 'failed',
        lastError: 'embedder timeout after 30s',
        retryCount: 3,
      })
      await stateStore.set(entry)

      const retrieved = await stateStore.get(entry.documentHash)
      expect(retrieved).toBeDefined()
      expect(retrieved!.lastError).toBe('embedder timeout after 30s')
      expect(retrieved!.retryCount).toBe(3)
    })

    it('preserves completedAt field', async () => {
      const completedAt = new Date('2026-01-05T12:00:00Z')
      const entry = makeEntry({ stage: 'completed', completedAt })
      await stateStore.set(entry)

      const retrieved = await stateStore.get(entry.documentHash)
      expect(retrieved).toBeDefined()
      expect(retrieved!.completedAt).toBeDefined()
      expect(retrieved!.completedAt!.toISOString()).toBe(completedAt.toISOString())
    })
  })
}

describe('FilePipelineStateStore prefix configuration', () => {
  it('uses a custom prefix when provided via options', async () => {
    const store: Store = createMemStore()
    const customStore = new FilePipelineStateStore({ store, prefix: 'raw/.custom-state-dir' })
    const defaultStore = new FilePipelineStateStore(store)

    const entry = makeEntry()
    await customStore.set(entry)

    const fromCustom = await customStore.get(entry.documentHash)
    expect(fromCustom).toBeDefined()
    expect(fromCustom!.documentHash).toBe(entry.documentHash)

    const fromDefault = await defaultStore.get(entry.documentHash)
    expect(fromDefault).toBeUndefined()

    await store.close()
  })

  it('uses the default prefix when prefix is omitted in options', async () => {
    const store: Store = createMemStore()
    const configStore = new FilePipelineStateStore({ store })
    const defaultStore = new FilePipelineStateStore(store)

    const entry = makeEntry()
    await configStore.set(entry)

    const fromDefault = await defaultStore.get(entry.documentHash)
    expect(fromDefault).toBeDefined()

    await store.close()
  })
})

describe('FilePipelineStateStore document hash validation', () => {
  let store: Store
  let stateStore: PipelineStateStore

  beforeEach(() => {
    store = createMemStore()
    stateStore = new FilePipelineStateStore(store)
  })

  afterEach(async () => {
    await store.close()
  })

  const invalidHashes = [
    '../../../etc/passwd',
    'short',
    'ABCD000000000000000000000000000000000000000000000000000000000000',
    'abc123def456',
    'abc/123/def/456',
  ]

  for (const hash of invalidHashes) {
    it(`rejects invalid hash "${hash}" on get`, async () => {
      await expect(stateStore.get(hash)).rejects.toThrow('Invalid document hash')
    })

    it(`rejects invalid hash "${hash}" on delete`, async () => {
      await expect(stateStore.delete(hash)).rejects.toThrow('Invalid document hash')
    })

    it(`rejects invalid hash "${hash}" on set`, async () => {
      const entry = makeEntry({ documentHash: hash })
      await expect(stateStore.set(entry)).rejects.toThrow('Invalid document hash')
    })
  }
})

File: go/ingest/state_store_test.go
// SPDX-License-Identifier: Apache-2.0
package ingest_test

import (
	"context"
	"testing"
	"time"

	"github.com/jeffs-brain/memory/go/ingest"
	"github.com/jeffs-brain/memory/go/store/fs"
)

const testBrainID = "test-brain"

// defaultHash is a valid 64-character lowercase hex string (SHA-256).
const defaultHash = "abc123def4560000000000000000000000000000000000000000000000000000"

func makeEntry(overrides ...func(*ingest.PipelineStateEntry)) ingest.PipelineStateEntry {
	entry := ingest.PipelineStateEntry{
		DocumentHash: defaultHash,
		BrainID:      testBrainID,
		Stage:        ingest.StageStored,
		RetryCount:   0,
		CreatedAt:    time.Date(2026, 1, 1, 0, 0, 0, 0, time.UTC),
		UpdatedAt:    time.Date(2026, 1, 1, 0, 0, 0, 0, time.UTC),
	}
	for _, fn := range overrides {
		fn(&entry)
	}
	return entry
}

// Factory creates a fresh PipelineStateStore for each subtest.
type Factory func(t *testing.T) ingest.PipelineStateStore

// runContractSuite executes the full contract test suite against any
// PipelineStateStore implementation.
func runContractSuite(t *testing.T, factory Factory) {
	t.Helper()

	cases := []struct {
		name string
		fn   func(t *testing.T, store ingest.PipelineStateStore)
	}{
		{"Get returns nil when no state exists", testGetNonExistent},
		{"Set then Get round-trips", testSetGetRoundTrip},
		{"Set overwrites existing entry", testSetOverwrite},
		{"ListIncomplete excludes completed and failed", testListIncompleteExcludes},
		{"ListIncomplete filters by brainID", testListIncompleteFiltersBrain},
		{"Delete removes entry", testDelete},
		{"Delete non-existent does not error", testDeleteNonExistent},
		{"Preserves lastError field", testPreservesLastError},
		{"Preserves completedAt field", testPreservesCompletedAt},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			store := factory(t)
			tc.fn(t, store)
		})
	}
}

func testGetNonExistent(t *testing.T, store ingest.PipelineStateStore) {
	t.Helper()
	ctx := context.Background()
	entry, err := store.Get(ctx, "0000000000000000000000000000000000000000000000000000000000000000")
	if err != nil {
		t.Fatalf("Get returned unexpected error: %v", err)
	}
	if entry != nil {
		t.Fatalf("expected nil entry, got %+v", entry)
	}
}

func testSetGetRoundTrip(t *testing.T, store ingest.PipelineStateStore) {
	t.Helper()
	ctx := context.Background()
	entry := makeEntry()

	if err := store.Set(ctx, entry); err != nil {
		t.Fatalf("Set: %v", err)
	}
	got, err := store.Get(ctx, entry.DocumentHash)
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	if got == nil {
		t.Fatal("Get returned nil after Set")
	}
	if got.DocumentHash != entry.DocumentHash {
		t.Errorf("DocumentHash = %q, want %q", got.DocumentHash, entry.DocumentHash)
	}
	if got.BrainID != entry.BrainID {
		t.Errorf("BrainID = %q, want %q", got.BrainID, entry.BrainID)
	}
	if got.Stage != entry.Stage {
		t.Errorf("Stage = %q, want %q", got.Stage, entry.Stage)
	}
	if got.RetryCount != entry.RetryCount {
		t.Errorf("RetryCount = %d, want %d", got.RetryCount, entry.RetryCount)
	}
	if !got.CreatedAt.Equal(entry.CreatedAt) {
		t.Errorf("CreatedAt = %v, want %v", got.CreatedAt, entry.CreatedAt)
	}
	if !got.UpdatedAt.Equal(entry.UpdatedAt) {
		t.Errorf("UpdatedAt = %v, want %v", got.UpdatedAt, entry.UpdatedAt)
	}
}

func testSetOverwrite(t *testing.T, store ingest.PipelineStateStore) {
	t.Helper()
	ctx := context.Background()
	entry := makeEntry()

	if err := store.Set(ctx, entry); err != nil {
		t.Fatalf("Set: %v", err)
	}

	updated := makeEntry(func(e *ingest.PipelineStateEntry) {
		e.Stage = ingest.StageChunked
		e.UpdatedAt = time.Date(2026, 1, 2, 0, 0, 0, 0, time.UTC)
	})
	if err := store.Set(ctx, updated); err != nil {
		t.Fatalf("Set overwrite: %v", err)
	}

	got, err := store.Get(ctx, entry.DocumentHash)
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	if got == nil {
		t.Fatal("Get returned nil after overwrite")
	}
	if got.Stage != ingest.StageChunked {
		t.Errorf("Stage = %q, want %q", got.Stage, ingest.StageChunked)
	}
	if !got.UpdatedAt.Equal(time.Date(2026, 1, 2, 0, 0, 0, 0, time.UTC)) {
		t.Errorf("UpdatedAt not updated: %v", got.UpdatedAt)
	}
}

func testListIncompleteExcludes(t *testing.T, store ingest.PipelineStateStore) {
	t.Helper()
	ctx := context.Background()

	hashReceived := "1111111111111111111111111111111111111111111111111111111111111111"
	hashStored := "2222222222222222222222222222222222222222222222222222222222222222"
	hashCompleted := "3333333333333333333333333333333333333333333333333333333333333333"
	hashFailed := "4444444444444444444444444444444444444444444444444444444444444444"
	hashEmbedded := "5555555555555555555555555555555555555555555555555555555555555555"

	entries := []ingest.PipelineStateEntry{
		makeEntry(func(e *ingest.PipelineStateEntry) { e.DocumentHash = hashReceived; e.Stage = ingest.StageReceived }),
		makeEntry(func(e *ingest.PipelineStateEntry) { e.DocumentHash = hashStored; e.Stage = ingest.StageStored }),
		makeEntry(func(e *ingest.PipelineStateEntry) { e.DocumentHash = hashCompleted; e.Stage = ingest.StageCompleted }),
		makeEntry(func(e *ingest.PipelineStateEntry) { e.DocumentHash = hashFailed; e.Stage = ingest.StageFailed }),
		makeEntry(func(e *ingest.PipelineStateEntry) { e.DocumentHash = hashEmbedded; e.Stage = ingest.StageEmbedded }),
	}
	for _, entry := range entries {
		if err := store.Set(ctx, entry); err != nil {
			t.Fatalf("Set %s: %v", entry.DocumentHash, err)
		}
	}

	incomplete, err := store.ListIncomplete(ctx, testBrainID)
	if err != nil {
		t.Fatalf("ListIncomplete: %v", err)
	}

	hashes := make(map[string]bool, len(incomplete))
	for _, e := range incomplete {
		hashes[e.DocumentHash] = true
	}

	if !hashes[hashReceived] {
		t.Error("expected hashReceived in incomplete list")
	}
	if !hashes[hashStored] {
		t.Error("expected hashStored in incomplete list")
	}
	if !hashes[hashEmbedded] {
		t.Error("expected hashEmbedded in incomplete list")
	}
	if hashes[hashCompleted] {
		t.Error("hashCompleted should not be in incomplete list")
	}
	if hashes[hashFailed] {
		t.Error("hashFailed should not be in incomplete list")
	}
	if len(incomplete) != 3 {
		t.Errorf("expected 3 incomplete entries, got %d", len(incomplete))
	}
}

func testListIncompleteFiltersBrain(t *testing.T, store ingest.PipelineStateStore) {
	t.Helper()
	ctx := context.Background()

	hashOurs := "aaaa000000000000000000000000000000000000000000000000000000000000"
	hashTheirs := "bbbb000000000000000000000000000000000000000000000000000000000000"

	ours := makeEntry(func(e *ingest.PipelineStateEntry) {
		e.DocumentHash = hashOurs
		e.BrainID = testBrainID
		e.Stage = ingest.StageStored
	})
	theirs := makeEntry(func(e *ingest.PipelineStateEntry) {
		e.DocumentHash = hashTheirs
		e.BrainID = "other-brain"
		e.Stage = ingest.StageStored
	})

	if err := store.Set(ctx, ours); err != nil {
		t.Fatalf("Set ours: %v", err)
	}
	if err := store.Set(ctx, theirs); err != nil {
		t.Fatalf("Set theirs: %v", err)
	}

	incomplete, err := store.ListIncomplete(ctx, testBrainID)
	if err != nil {
		t.Fatalf("ListIncomplete: %v", err)
	}
	if len(incomplete) != 1 {
		t.Fatalf("expected 1 entry, got %d", len(incomplete))
	}
	if incomplete[0].DocumentHash != hashOurs {
		t.Errorf("expected %s, got %s", hashOurs, incomplete[0].DocumentHash)
	}
}

func testDelete(t *testing.T, store ingest.PipelineStateStore) {
	t.Helper()
	ctx := context.Background()
	entry := makeEntry()

	if err := store.Set(ctx, entry); err != nil {
		t.Fatalf("Set: %v", err)
	}
	if err := store.Delete(ctx, entry.DocumentHash); err != nil {
		t.Fatalf("Delete: %v", err)
	}

	got, err := store.Get(ctx, entry.DocumentHash)
	if err != nil {
		t.Fatalf("Get after Delete: %v", err)
	}
	if got != nil {
		t.Fatalf("expected nil after Delete, got %+v", got)
	}
}

func testDeleteNonExistent(t *testing.T, store ingest.PipelineStateStore) {
	t.Helper()
	ctx := context.Background()
	if err := store.Delete(ctx, "0000000000000000000000000000000000000000000000000000000000000000"); err != nil {
		t.Fatalf("Delete non-existent returned error: %v", err)
	}
}

func testPreservesLastError(t *testing.T, store ingest.PipelineStateStore) {
	t.Helper()
	ctx := context.Background()
	entry := makeEntry(func(e *ingest.PipelineStateEntry) {
		e.Stage = ingest.StageFailed
		e.LastError = "embedder timeout after 30s"
		e.RetryCount = 3
	})

	if err := store.Set(ctx, entry); err != nil {
		t.Fatalf("Set: %v", err)
	}
	got, err := store.Get(ctx, entry.DocumentHash)
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	if got == nil {
		t.Fatal("Get returned nil")
	}
	if got.LastError != "embedder timeout after 30s" {
		t.Errorf("LastError = %q, want %q", got.LastError, "embedder timeout after 30s")
	}
	if got.RetryCount != 3 {
		t.Errorf("RetryCount = %d, want 3", got.RetryCount)
	}
}

func testPreservesCompletedAt(t *testing.T, store ingest.PipelineStateStore) {
	t.Helper()
	ctx := context.Background()
	completedAt := time.Date(2026, 1, 5, 12, 0, 0, 0, time.UTC)
	entry := makeEntry(func(e *ingest.PipelineStateEntry) {
		e.Stage = ingest.StageCompleted
		e.CompletedAt = &completedAt
	})

	if err := store.Set(ctx, entry); err != nil {
		t.Fatalf("Set: %v", err)
	}
	got, err := store.Get(ctx, entry.DocumentHash)
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	if got == nil {
		t.Fatal("Get returned nil")
	}
	if got.CompletedAt == nil {
		t.Fatal("CompletedAt is nil after round-trip")
	}
	if !got.CompletedAt.Equal(completedAt) {
		t.Errorf("CompletedAt = %v, want %v", got.CompletedAt, completedAt)
	}
}

// TestFilePipelineStateStore runs the full contract suite against the
// file-based implementation using a real filesystem-backed brain.Store.
func TestFilePipelineStateStore(t *testing.T) {
	runContractSuite(t, func(t *testing.T) ingest.PipelineStateStore {
		t.Helper()
		store, err := fs.New(t.TempDir())
		if err != nil {
			t.Fatalf("fs.New: %v", err)
		}
		t.Cleanup(func() { _ = store.Close() })
		return ingest.NewFilePipelineStateStore(store)
	})
}

// TestFilePipelineStateStore_CustomPrefix verifies that a custom path prefix
// is honoured by the state store.
func TestFilePipelineStateStore_CustomPrefix(t *testing.T) {
	store, err := fs.New(t.TempDir())
	if err != nil {
		t.Fatalf("fs.New: %v", err)
	}
	t.Cleanup(func() { _ = store.Close() })

	customPrefix := "raw/.custom-state-dir"
	ss := ingest.NewFilePipelineStateStoreWithConfig(ingest.FilePipelineStateStoreConfig{
		Store:  store,
		Prefix: customPrefix,
	})

	ctx := context.Background()
	entry := makeEntry()

	if err := ss.Set(ctx, entry); err != nil {
		t.Fatalf("Set: %v", err)
	}

	got, err := ss.Get(ctx, entry.DocumentHash)
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	if got == nil {
		t.Fatal("Get returned nil after Set with custom prefix")
	}
	if got.DocumentHash != entry.DocumentHash {
		t.Errorf("DocumentHash = %q, want %q", got.DocumentHash, entry.DocumentHash)
	}

	// Verify the file is stored under the custom prefix, not the default.
	defaultStore := ingest.NewFilePipelineStateStore(store)
	fromDefault, err := defaultStore.Get(ctx, entry.DocumentHash)
	if err != nil {
		t.Fatalf("Get from default store: %v", err)
	}
	if fromDefault != nil {
		t.Error("entry should not be found under the default prefix")
	}
}

// TestFilePipelineStateStore_DefaultPrefix verifies that omitting the
// prefix in config uses the default "raw/.pipeline-state".
func TestFilePipelineStateStore_DefaultPrefix(t *testing.T) {
	store, err := fs.New(t.TempDir())
	if err != nil {
		t.Fatalf("fs.New: %v", err)
	}
	t.Cleanup(func() { _ = store.Close() })

	ssConfig := ingest.NewFilePipelineStateStoreWithConfig(ingest.FilePipelineStateStoreConfig{
		Store: store,
	})
	ssDefault := ingest.NewFilePipelineStateStore(store)

	ctx := context.Background()
	entry := makeEntry()

	if err := ssConfig.Set(ctx, entry); err != nil {
		t.Fatalf("Set via config: %v", err)
	}

	got, err := ssDefault.Get(ctx, entry.DocumentHash)
	if err != nil {
		t.Fatalf("Get via default: %v", err)
	}
	if got == nil {
		t.Fatal("default prefix should find entry written with empty config prefix")
	}
}

// TestFilePipelineStateStore_RejectsInvalidHash verifies that path-traversal
// or otherwise invalid document hashes are rejected.
func TestFilePipelineStateStore_RejectsInvalidHash(t *testing.T) {
	store, err := fs.New(t.TempDir())
	if err != nil {
		t.Fatalf("fs.New: %v", err)
	}
	t.Cleanup(func() { _ = store.Close() })
	ss := ingest.NewFilePipelineStateStore(store)

	ctx := context.Background()
	invalidHashes := []string{
		"../../../etc/passwd",
		"short",
		"ABCD000000000000000000000000000000000000000000000000000000000000", // uppercase
		"abc123def456",    // too short
		"abc/123/def/456", // path separators
	}

	for _, hash := range invalidHashes {
		_, err := ss.Get(ctx, hash)
		if err == nil {
			t.Errorf("Get(%q) should have returned error for invalid hash", hash)
		}

		err = ss.Delete(ctx, hash)
		if err == nil {
			t.Errorf("Delete(%q) should have returned error for invalid hash", hash)
		}

		err = ss.Set(ctx, ingest.PipelineStateEntry{
			DocumentHash: hash,
			BrainID:      testBrainID,
			Stage:        ingest.StageStored,
		})
		if err == nil {
			t.Errorf("Set with hash %q should have returned error for invalid hash", hash)
		}
	}
}

// TestPostgresPipelineStateStore_RejectsNilDB verifies that the constructor
// rejects a nil DB connection.
func TestPostgresPipelineStateStore_RejectsNilDB(t *testing.T) {
	_, err := ingest.NewPostgresPipelineStateStore(ingest.PostgresStateStoreConfig{
		DB: nil,
	})
	if err == nil {
		t.Fatal("expected error for nil DB, got nil")
	}
}
